eth/chain.go

```diff
@@ -0,0 +1,455 @@
+	oldEpoch := oldSt.FinalizedEpoch
+	oldFork := c.Fork
+
+	// compute new status
+	currentSlot := slots.CurrentSlot(c.cfg.GenesisConfig.GenesisTime)
+	currentEpoch := slots.ToEpoch(currentSlot)
+	currentForkDigest := params.ForkDigest(currentEpoch)
+	currentFork := c.forkFromEpoch(currentEpoch)
+	c.statusMu.Lock()
+	c.Fork = currentFork
+	slog.Info(
+		"chain set to",
+		"fork", c.Fork,
+		"head-slot", currentSlot,
+	)
+	c.statusMu.Unlock()
+
+	// recompute the status
+	newSt := &pb.StatusV2{
+		ForkDigest:            currentForkDigest[:],
+		FinalizedRoot:         chainHead.FinalizedBlockRoot,
+		FinalizedEpoch:        chainHead.FinalizedEpoch,
+		HeadRoot:              chainHead.HeadBlockRoot,
+		HeadSlot:              chainHead.HeadSlot,
+		EarliestAvailableSlot: chainHead.HeadSlot,
+	}
+	c.UpdateStatus(statusV1, newSt)
+
+	// recompute the metadata (if needed) - for now, no changes are expected
+
+	// notify of any update to whoever needs it
+	notifyNewFork := (currentEpoch > oldEpoch && currentFork > oldFork)
+	if notifyNewFork {
+		for _, notFn := range c.chainUpgradeSubs {
+			if err := notFn(); err != nil {
+				slog.Error(
+					"error notifying sub of new fork",
+					"error", err.Error(),
+				)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *Chain) epochStats() (primitives.Slot, primitives.Epoch, chainFork, [4]byte, error) {
+	if c.cfg.BeaconConfig == nil || c.cfg.GenesisConfig == nil {
+		return 0, 0, phase0, [4]byte{}, fmt.Errorf("chain internals doest have beacon config")
+	}
+	var (
+		slot       = slots.CurrentSlot(c.cfg.GenesisConfig.GenesisTime)
+		epoch      = slots.ToEpoch(slot)
+		forkDigest = params.ForkDigest(epoch)
+	)
+
+	return slot, epoch, c.forkFromEpoch(epoch), forkDigest, nil
+}
+
+func (c *Chain) forkFromEpoch(epoch primitives.Epoch) chainFork {
+	var fork = phase0
+	if c.cfg.BeaconConfig.FuluForkEpoch != params.BeaconConfig().FarFutureEpoch && epoch >= c.cfg.BeaconConfig.FuluForkEpoch {
+		fork = fulu
+	} else if c.cfg.BeaconConfig.ElectraForkEpoch != params.BeaconConfig().FarFutureEpoch && epoch >= c.cfg.BeaconConfig.ElectraForkEpoch {
+		fork = electra
+	} else if c.cfg.BeaconConfig.DenebForkEpoch != params.BeaconConfig().FarFutureEpoch && epoch >= c.cfg.BeaconConfig.DenebForkEpoch {
+		fork = deneb
+	} else if c.cfg.BeaconConfig.CapellaForkEpoch != params.BeaconConfig().FarFutureEpoch && epoch >= c.cfg.BeaconConfig.CapellaForkEpoch {
+		fork = capella
+	} else if c.cfg.BeaconConfig.BellatrixForkEpoch != params.BeaconConfig().FarFutureEpoch && epoch >= c.cfg.BeaconConfig.BellatrixForkEpoch {
+		fork = bellatrix
+	} else if c.cfg.BeaconConfig.AltairForkEpoch != params.BeaconConfig().FarFutureEpoch && epoch >= c.cfg.BeaconConfig.AltairForkEpoch {
+		fork = altair
+	}
+	return fork
+}
+
+func (c *Chain) Serve(ctx context.Context) error {
+	// launch internal loop
+	defer func() {
+		slog.Info("chain internal loop closing up")
+	}()
+	for {
+		select {
+		// Epoch iteration:
+		case <-time.After(384 * time.Second):
```

Comment:
This looks to me as it could get progressively out of sync. Can the time.After duration be computed from the result of the epoch Update? I mean, let's say the `epochUpdate` call takes 2s, then we're slowly drifting or is that not an issue?